### PR TITLE
thrift/inbound: Track application errors

### LINKED ISSUE #163: Thrift exceptions should use flags=0x01
Per the [Thrift over TChannel](https://github.com/uber/tchannel/blob/master/docs/thrift.md) Protocol doc - 

```
In case of failure, the Response Code (code:1) must be set to 0x01.
```

Instead, we're setting `0x00` - 

```
$ sudo tcap -i lo0 -i en0 -p 8080 -p 8082
...

ts=1464736322.907 session=4 127.0.0.1:64892 <-- 127.0.0.1:8082 frame=2 type=0x04 Ok
CALL RESPONSE id=0x0002 (2) flags=0x00
headers
  as: thrift
tracing: spanid=0,0 parentid=0,0 traceid=0,0 flags=0x00
args[0]
    00:                                          empty
args[1]
    00: 0000                                     ..
args[2]
    00: 0c00 0108 0001 0000 03e9 0b00 0200 0000  ................
    10: 0858 6365 7074 696f 6e00 00              .Xception..
arg3 as thrift
{ '1':
   { '1': 1001,
     '2': 'Xception' } }
```

This is blocking #162's `TestException` test since the error is being set to  `nil` instead of `gauntlet_apache.Xception`.

cc @yarpc/golang

### PR DESCRIPTION
**Pull request by abhinav**
This uses the new ResponseWriter API to inform the transport that the response
body contains an application error, which allows TChannel to set the error bit
in the Call Response.

This also enables the Thrift application exception tests in the tchclient behavior.

Related code generation change in thriftrw-go: thriftrw/thriftrw-go#149

Resolves #163

CC @yarpc/golang

### CODE REVIEW

File: encoding/thrift/inbound_test.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thriftrw/thriftrw-go/wire"
+	"golang.org/x/net/context"
+)
+
+//go:generate mockgen -destination=mock_handler_test.go -package=thrift github.com/yarpc/yarpc-go/encoding/thrift Handler
+//go:generate mockgen -destination=mock_protocol_test.go -package=thrift github.com/thriftrw/thriftrw-go/protocol Protocol
```
Comment (breerly) [inline encoding/thrift/inbound_test.go]: first time I've seen these not in `gen.go` - guessing thats coo?

File: encoding/thrift/register.go [hunk 2]
```diff
@@ -27,16 +27,23 @@ import (
 	"github.com/thriftrw/thriftrw-go/wire"
 )
 
+// TResponse contains the response from a generated Thrift handler.
+type TResponse struct { // TODO(abg): Rename to Response
```
Comment (breerly) [inline encoding/thrift/register.go]: Lets capture a ticket for this one - important enough forward-facing change.

File: encoding/thrift/inbound_test.go (hunk 1 above)
Comment (abhinav) [inline encoding/thrift/inbound_test.go]: Yeah this is fine. The generated code is needed for the tests in this file only so there's no need to have the generate clauses elsewhere.

File: encoding/thrift/inbound.go [hunk 3]
```diff
@@ -61,11 +61,16 @@ func (t thriftHandler) Handle(ctx context.Context, treq *transport.Request, rw t
 		return err
 	}
 
+	if res.IsApplicationError {
+		rw.SetApplicationError()
```
Comment (breerly) [inline encoding/thrift/inbound.go]: nice, this is really clean

**Comment by breerly** [comment]
lgtm 💯 💯 💯

File: crossdock/thrift/echo/yarpc/echoserver/server.go [hunk 4]
```diff
@@ -54,16 +54,19 @@ func (s service) Handlers() map[string]thrift.Handler {
...
 	var args echo2.EchoArgs
 	if err := args.FromWire(body); err != nil {
-		return wire.Value{}, nil, err
+		return thrift.TResponse{}, err
 	}
 	success, res, err := h.impl.Echo(req, args.Ping)
+	hadError := err != nil
 	result, err := echo2.EchoHelper.WrapResponse(success, err)
-	var w wire.Value
+	var response thrift.TResponse
```
Comment (breerly) [inline crossdock/thrift/echo/yarpc/echoserver/server.go]: Nice, so `TResponse` is never exposed directly to the user 👍

**Comment by prashantv** [comment]
lgtm

**Pull request merged**